utils/hparams.py

```diff
@@ -101,30 +101,37 @@ def load_config(config_fn):  # deep first
             else:
                 hparams_[k] = type(hparams_[k])(v)
 
-    if args_work_dir != '' and (not os.path.exists(ckpt_config_path) or args.reset) and not args.infer:
-        os.makedirs(hparams_['work_dir'], exist_ok=True)
-        if is_main_process:
-            # Only the main process will save the config file
-            with open(ckpt_config_path, 'w', encoding='utf-8') as f:
-                hparams_non_recursive = hparams_.copy()
-                hparams_non_recursive['base_config'] = []
-                yaml.safe_dump(hparams_non_recursive, f, allow_unicode=True, encoding='utf-8')
+    @rank_zero_only
+    def dump_hparams():
+        if args_work_dir != '' and (not os.path.exists(ckpt_config_path) or args.reset) and not args.infer:
+            os.makedirs(hparams_['work_dir'], exist_ok=True)
+            if mp_is_main_process:
+                # Only the main process will save the config file
+                with open(ckpt_config_path, 'w', encoding='utf-8') as f:
+                    hparams_non_recursive = hparams_.copy()
+                    hparams_non_recursive['base_config'] = []
+                    yaml.safe_dump(hparams_non_recursive, f, allow_unicode=True, encoding='utf-8')
+    dump_hparams()
 
     hparams_['infer'] = args.infer
     hparams_['debug'] = args.debug
     hparams_['validate'] = args.validate
-    global global_print_hparams
     if global_hparams:
         hparams.clear()
         hparams.update(hparams_)
-
-    if is_main_process and print_hparams and global_print_hparams and global_hparams:
-        print('| Hparams chains: ', config_chains)
-        print('| Hparams: ')
-        for i, (k, v) in enumerate(sorted(hparams_.items())):
-            print(f"\033[;33;m{k}\033[0m: {v}, ", end="\n" if i % 5 == 4 else "")
-        print("")
-        global_print_hparams = False
+    
+    @rank_zero_only
+    def print_hparams():
+        global global_print_hparams
+        if mp_is_main_process and print_hparams and global_print_hparams and global_hparams:
+            print('| Hparams chains: ', config_chains)
+            print('| Hparams: ')
+            for i, (k, v) in enumerate(sorted(hparams_.items())):
+                print(f"\033[;33;m{k}\033[0m: {v}, ", end="\n" if i % 5 == 4 else "")
+            print("")
+            global_print_hparams = False
+    print_hparams()
```

Comment:
`print_hparams()` call should be below all changes to `hparams`. Here `exp_name` is set after this call.